streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java

```diff
@@ -570,13 +591,34 @@ public StreamsConfig(final Map<?, ?> props) {
      * @return Map of the producer configuration.
      */
     public Map<String, Object> getProducerConfigs(final String clientId) {
+        final Map<String, Object> clientProvidedProps = getClientPropsWithPrefix(PRODUCER_PREFIX, ProducerConfig.configNames());
+
         // generate producer configs from original properties and overridden maps
         final Map<String, Object> props = new HashMap<>(PRODUCER_DEFAULT_OVERRIDES);
         props.putAll(getClientPropsWithPrefix(PRODUCER_PREFIX, ProducerConfig.configNames()));
 
         props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, originals().get(BOOTSTRAP_SERVERS_CONFIG));
         // add client id with stream client id prefix
         props.put(CommonClientConfigs.CLIENT_ID_CONFIG, clientId + "-producer");
+        if (eosEnabled) {
+            if (clientProvidedProps.containsKey(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG)) {
+                throw new ConfigException("Unexpected user-specified consumer config " + ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG
+                    + "; because " + PROCESSING_GUARANTEE_CONFIG + " is set to '" + EXACTLY_ONCE + "' producer will always have idempotency enabled.");
+            }
+            props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+
+            if (clientProvidedProps.containsKey(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION)) {
+                throw new ConfigException("Unexpected user-specified consumer config " + ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION
```

Comment:
The numbers are here: https://docs.google.com/spreadsheets/d/1dHY6M7qCiX-NFvsgvaE0YoVdNq26uA8608XIh_DUpI4/edit#gid=1276994626

As you can see ,setting inflights to 1 from the default of 5 does have a perf impact of 20% on those tests. But also note that compression was enabled and batch sizes were small.  

I think that if you tune your batches to be larger, the impact will be ameliorated. For instance, if you commit transactions every 500ms, your linger ms could be 500ms as well for maximum batches. With larger batches, you don't get much benefit from multiple inflights.

Finally, we can't guarantee in order delivery without this setting even before KIP-98. Are all streams operators currently symmetric? Otherwise they would already be buggy.